### ISSUE TITLE
[BUG] Computation time of MAE (and possibly other metrics) is very high

### ISSUE THREAD

**Issue by AlexMV12** [unit8co/darts#2249]
**Describe the bug**
The computation time of MAE and MSE (i..e, `from darts.metrics import mae`) is very high w.r.t. `sklearn` ones, especially when time-series contain many features.

**To Reproduce**
You can change the parameters `N_FEATURES` and `N_EXPERIMENTS` to get different results.

```python
from darts import TimeSeries
from darts.metrics import mae as mae_function_darts
import time
from sklearn.metrics import mean_absolute_error as mae_function_sklearn

import numpy as np
import pandas as pd

# Set the seed for reproducibility
np.random.seed(42)

# We want to check the difference in computation time between
# the MAE computed by Darts, and the MAE computed using sklearn.

N_FEATURES = 300
N_EXPERIMENTS = 100

for ts_length in [100, 1000, 10000]:
    print(f"Testing with ts length {ts_length}")

    # Create two similar but slightly different datasets.
    data1 = np.random.randint(1, 10, size=(ts_length, N_FEATURES))
    data2 = data1 + np.random.randint(-2, 3, size=(ts_length, N_FEATURES))  # Slight variation

    # Create a date range for the time-series index
    dates = pd.date_range(start='2022-01-01', periods=ts_length, freq='1min')

    # Create Darts TimeSeries objects
    ts1 = TimeSeries.from_dataframe(pd.DataFrame(data1, columns=[str(x) for x in range(0, N_FEATURES)], index=dates))
    ts2 = TimeSeries.from_dataframe(pd.DataFrame(data2, columns=[str(x) for x in range(0, N_FEATURES)], index=dates))

    # Now, compute the MAE and MSE using both sklearn and darts.
    darts_times = []
    sklearn_times = []

    for experiment in range(N_EXPERIMENTS):
        start_darts = time.time()
        mae_darts = mae_function_darts(ts1, ts2)
        darts_times.append(time.time() - start_darts)

        # Wait...
        time.sleep(1)

        start_sklearn = time.time()
        mae_sklearn = mae_function_sklearn(ts1.values(), ts2.values())
        sklearn_times.append(time.time() - start_sklearn)

        assert mae_darts == mae_sklearn
    
    print(f"Time requested by Darts (mean +- std): {np.mean(darts_times):.5f} +- {np.std(darts_times):.5f}")
    print(f"Time requested by sklearn (mean +- std): {np.mean(sklearn_times):.5f} +- {np.std(sklearn_times):.5f}")

    # Wait...
    time.sleep(1)


```

**Expected behavior**
The time requested to compute the MAE should not be so much different from the one of `sklearn`. This is especially problematic when computing historical forecasts on big datasets (e.g., `ETTm1` and `ETTm2`, or `Electricity`) which can contain hundreds of features, and hunders/thousands of predictions. It is even slower to compute the MAE on the predictions that the predictions themselves (for certain models)!

**System (please complete the following information):**
 - Python version: 3.11.6
 - darts version: darts-0.27.2

**Additional context**
- This happens to me on several machines; one with ArchLinux, one with Ubuntu, even on Docker (in that case, just pull the `python` image, install `darts`, and run the test).
- This happens also with the MSE; I don't know if it happens with every metric, but it may be worth investigating.
- It is true that Darts also looks at the time-index to make sure that each points is checked with the corresponding right point in `y_true`, but even with that, it looks a bit too slow.

These are some runs, if you want to get an idea of how bad it is:

```
N_FEATURES = 3
N_EXPERIMENTS = 10

Testing with ts length 100
Time requested by Darts (mean +- std): 0.04003 +- 0.00718
Time requested by sklearn (mean +- std): 0.00191 +- 0.00009
Testing with ts length 1000
Time requested by Darts (mean +- std): 0.03791 +- 0.00574
Time requested by sklearn (mean +- std): 0.00201 +- 0.00007
Testing with ts length 10000
Time requested by Darts (mean +- std): 0.03962 +- 0.00443
Time requested by sklearn (mean +- std): 0.00283 +- 0.00004
```

```
N_FEATURES = 30
N_EXPERIMENTS = 10

Testing with ts length 100
Time requested by Darts (mean +- std): 0.23278 +- 0.00389
Time requested by sklearn (mean +- std): 0.00200 +- 0.00013
Testing with ts length 1000
Time requested by Darts (mean +- std): 0.21415 +- 0.00982
Time requested by sklearn (mean +- std): 0.00245 +- 0.00013
Testing with ts length 10000
Time requested by Darts (mean +- std): 0.24863 +- 0.00285
Time requested by sklearn (mean +- std): 0.01092 +- 0.00010
```

```
N_FEATURES = 300
N_EXPERIMENTS = 10

Testing with ts length 100
Time requested by Darts (mean +- std): 1.55213 +- 0.23301
Time requested by sklearn (mean +- std): 0.00247 +- 0.00029
Testing with ts length 1000
Time requested by Darts (mean +- std): 1.29242 +- 0.01550
Time requested by sklearn (mean +- std): 0.01023 +- 0.00019
Testing with ts length 10000
Time requested by Darts (mean +- std): 1.62117 +- 0.17096
Time requested by sklearn (mean +- std): 0.03871 +- 0.00080
```

```
N_FEATURES = 1000
N_EXPERIMENTS = 10

Testing with ts length 100
Time requested by Darts (mean +- std): 4.87219 +- 0.20418
Time requested by sklearn (mean +- std): 0.00352 +- 0.00053
Testing with ts length 1000
Time requested by Darts (mean +- std): 4.31733 +- 0.01177
Time requested by sklearn (mean +- std): 0.01945 +- 0.00058
Testing with ts length 10000
Time requested by Darts (mean +- std): 5.24525 +- 0.01451
Time requested by sklearn (mean +- std): 0.12486 +- 0.00085
```

**Comment by dennisbader** [comment]
Thank you @AlexMV12 for this well documented issue.

Indeed there is potential for improving the metric efficiency. Is this something you'd like to contribute?

**Comment by dennisbader** [comment]
@AlexMV12, btw at the moment you can already speed things quite up a bit by computing the darts metric on a list of series (in parallel). Below is an adapted example which benchmarks the multiple series metric, against an sklearn metric on the concatenated y_true, y_actual values (giving sklearn a huge advantage, as they don't have to deal with the slicing).

This makes Darts ~20x faster than before.

It's not meant to deny that we definitely need to improve the efficiency, just that there are possibilities to speed things up :) 

```
from darts import TimeSeries
from darts.metrics import mae as mae_function_darts
import time
from sklearn.metrics import mean_absolute_error as mae_function_sklearn

import numpy as np
import pandas as pd

# Set the seed for reproducibility
np.random.seed(42)

# We want to check the difference in computation time between
# the MAE computed by Darts, and the MAE computed using sklearn.

N_FEATURES = 300
N_EXPERIMENTS = 100

times = {"darts": [], "sklearn": []}
for ts_length in [100, 1000, 10000]:
    print(f"Testing with ts length {ts_length}")

    # Create two similar but slightly different datasets.
    data1 = np.random.randint(1, 10, size=(ts_length, N_FEATURES))
    data2 = data1 + np.random.randint(-2, 3, size=(ts_length, N_FEATURES))  # Slight variation

    # Create a date range for the time-series index
    dates = pd.date_range(start='2022-01-01', periods=ts_length, freq='1min')

    # Create Darts TimeSeries objects
    ts1 = TimeSeries.from_dataframe(pd.DataFrame(data1, columns=[str(x) for x in range(0, N_FEATURES)], index=dates))
    ts2 = TimeSeries.from_dataframe(pd.DataFrame(data2, columns=[str(x) for x in range(0, N_FEATURES)], index=dates))

    # Now, compute the MAE and MSE using both sklearn and darts.
    start_darts = time.time()

    series_true, series_pred = [ts1] * N_EXPERIMENTS, [ts2] * N_EXPERIMENTS
    y_true = np.concatenate([s.values(copy=False) for s in series_true], axis=0)
    y_pred = np.concatenate([s.values(copy=False) for s in series_pred], axis=0)

    start_darts = time.time()
    mae_darts = np.mean(mae_function_darts([ts1] * N_EXPERIMENTS, [ts2] * N_EXPERIMENTS, verbose=True, n_jobs=-1))
    darts_time = time.time() - start_darts

    start_sklearn = time.time()
    mae_sklearn = mae_function_sklearn(y_true, y_pred)
    sklearn_time = time.time() - start_sklearn

    np.testing.assert_almost_equal(mae_darts, mae_sklearn)

    print(f"Time requested by Darts (mean +- std): {darts_time / N_EXPERIMENTS:.5f}")
    print(f"Time requested by sklearn (mean +- std): {sklearn_time / N_EXPERIMENTS:.5f}")
    times["darts"] = darts_time
    times["sklearn"] = sklearn_time
```

Results for `N_FEATURES = 300` (darts is 

```
N_FEATURES = 300
N_EXPERIMENTS = 100

Testing with ts length 100
100%|██████████| 100/100 [00:07<00:00, 14.21it/s]
Time requested by Darts (mean +- std): 0.07965
Time requested by sklearn (mean +- std): 0.00009
Testing with ts length 1000
100%|██████████| 100/100 [00:05<00:00, 19.00it/s]
Time requested by Darts (mean +- std): 0.06247
Time requested by sklearn (mean +- std): 0.00187
Testing with ts length 10000
100%|██████████| 100/100 [00:08<00:00, 11.38it/s]
Time requested by Darts (mean +- std): 0.10126
Time requested by sklearn (mean +- std): 0.04918
```

Results for `N_FEATURES = 1000`

```
N_FEATURES = 1000
N_EXPERIMENTS = 100

Testing with ts length 100
100%|██████████| 100/100 [00:20<00:00,  4.87it/s]
Time requested by Darts (mean +- std): 0.23028
Time requested by sklearn (mean +- std): 0.00037
Testing with ts length 1000
100%|██████████| 100/100 [00:17<00:00,  5.78it/s]
Time requested by Darts (mean +- std): 0.20000
Time requested by sklearn (mean +- std): 0.00583
Testing with ts length 10000
100%|██████████| 100/100 [00:31<00:00,  3.22it/s]
Time requested by Darts (mean +- std): 0.35540
Time requested by sklearn (mean +- std): 0.37899
```

**Comment by AlexMV12** [comment]
Hello @dennisbader ,

Thanks to you for Darts and for your answer!
I am not completely familiar with the inner working of Darts, but I would be happy to contribute in some way given that it is indeed a very good piece of software. I will try to give a look at how metrics are computed to see if I am able to bring something on the table.

And thanks for the idea on parallel computation!

**Comment by dennisbader** [comment]
Hi @AlexMV12, I'm also currently investigating this. Apart from the metric wrappers for multivariate and multiple time series support, there are also several `TimeSeries` methods which can be improved. This is pretty deep inside Darts, so I can also take over if you want :)

**Comment by AlexMV12** [comment]
Hi @dennisbader, sure, feel free to take over! :)
Thanks again for your support.
